### PR TITLE
feat(optimizer): Add `StreamProjectMergeRule`

### LINKED ISSUE #8742: feat: StreamProject Merge for two phase plans
Use this regex to search our testcases for these cases: `StreamProject.*\n.*StreamProject \{`.

Example:
https://github.com/risingwavelabs/risingwave/blob/96aa23dd0dc04c869a2ea5fdb4b7b1fc8fbcbf6d/src/frontend/planner_test/tests/testdata/agg.yaml#L1182-L1183

It occurs because `Project` is inserted (during `to_stream_dist`) to add `vnode`, after our optimization pass for `ProjectMergeRule` (during `gen_optimized_logical_plan_for_stream`.)

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Adds `StreamProjectMergeRule` which merges projects inserted during stream plan generation.
- Refactor optimization steps of stream plan into its own function: `gen_optimized_stream_plan`.
- Add sink planner test for nexmark queries. Not parameterized at the moment. Limited to blackhole sink. Required to test q0's duplicate projects, see query plan in #8712.

Closes #8742 
Workaround #8577 
Unblock eliminate `compact` in #8712 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/nexmark.yaml [hunk 1]
```diff
@@ -1751,13 +1751,12 @@
         └─StreamExchange { dist: Single }
           └─StreamGroupTopN { order: "[count(bid.auction) DESC]", limit: 1000, offset: 0, group_key: [3] }
             └─StreamProject { exprs: [auction.id, auction.item_name, count(bid.auction), Vnode(auction.id) as $expr1] }
-              └─StreamProject { exprs: [auction.id, auction.item_name, count(bid.auction)] }
```
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/nexmark.yaml:1754]: Merged into the one above

File: src/frontend/planner_test/tests/testdata/agg.yaml [hunk 2]
```diff
@@ -1180,14 +1178,13 @@
...
+              └─StreamHashAgg { group_key: [bid.auction, window_start], aggs: [sum0(count), count] }
+                └─StreamExchange { dist: HashShard(bid.auction, window_start) }
+                  └─StreamHashAgg { group_key: [bid.auction, window_start, $expr1], aggs: [count] }
+                    └─StreamProject { exprs: [bid.auction, window_start, bid._row_id, Vnode(bid._row_id) as $expr1] }
+                      └─StreamHopWindow { time_col: bid.date_time, slide: 00:00:02, size: 00:00:10, output: [bid.auction, window_start, bid._row_id] }
+                        └─StreamFilter { predicate: IsNotNull(bid.date_time) }
+                          └─StreamTableScan { table: bid, columns: [bid.date_time, bid.auction, bid._row_id], pk: [bid._row_id], dist: UpstreamHashShard(bid._row_id) }
 - name: two phase agg with stream SomeShard (via index) but pk satisfies output dist should use shuffle agg
   sql: |
     SET QUERY_MODE TO DISTRIBUTED;
```
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/agg.yaml:1190]: two phase agg projects are merged here. #8742

File: src/frontend/planner_test/tests/testdata/nexmark.yaml [hunk 3]
```diff
@@ -42,6 +42,11 @@
   batch_plan: |
     BatchExchange { order: [], dist: Single }
     └─BatchScan { table: bid, columns: [bid.auction, bid.bidder, bid.price, bid.date_time], distribution: SomeShard }
+  sink_plan: |
+    StreamSink { type: append-only, columns: [auction, bidder, price, date_time] }
+    └─StreamProject { exprs: [bid.auction, bid.bidder, bid.price, bid.date_time] }
```
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/nexmark.yaml:47]: Nexmark q0 fixed here cc @lmatz

**Review by lmatz — APPROVED** [review]

**Review by st1page — APPROVED** [review]
LGTM

**Comment by lmatz** [comment]
I found several times that buildkite is not triggered after turning a `draft` into `ready for review`...

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**